### DISCUSSION TITLE
Inconsistency in retrieving polygon data with get_polygons_points() for VInstance objects created via route_bundle_all_angle vs. create_vinst()

### DISCUSSION THREAD

**Discussion by Dangerance** [gdsfactory/gdsfactory#3500] [Q&A]
Hello gdsfactory developers,

I am encountering an issue when trying to retrieve polygon data with get_polygons_points() from VInstance objects that are generated differently. Specifically, instances created via create_vinst() work as expected, but instances resulting from route_bundle_all_angle() calls do not.

Problem Description:
Instances created using c.create_vinst(mmi) successfully return polygon data with inst.cell.get_polygons_points().

However, instances appearing after route_bundle_all_angle() calls raise errors, such as 'ComponentAllAngle' object has no attribute 'cell' and list index out of range, when attempting to invoke get_polygons_points().Calling inst.cell.get_polygons_points() on these routed instances leads to errors such as:
①'ComponentAllAngle' object has no attribute 'cell'
②list index out of range

These two sets of instances are both reported as <class 'kfactory.kcell.VInstance'>, yet they behave inconsistently. My main goal is to be able to consistently extract polygon data from all VInstance objects, including those generated by route_bundle_all_angle().

Additional Questions:
1、Could you clarify the difference in how route_bundle_all_angle handles or generates VInstance objects internally?
2、Is there a recommended approach to accessing polygon data from the OpticalAllAngleRoute objects returned by route_bundle_all_angle?

I am using the latest version of GDSFactory, with Python 3.12.
Code：
import gdsfactory as gf

c = gf.ComponentAllAngle()
mmi = gf.components.mmi1x2()

mmi1 = c.create_vinst(mmi)
mmi2 = c.create_vinst(mmi)
mmi3 = c.create_vinst(mmi)
mmi4 = c.create_vinst(mmi)
mmi5 = c.create_vinst(mmi)

mmi1.movey(50).movex(50).rotate(30)
mmi2.movey(100).movex(150).rotate(10)
mmi3.movey(100).movex(200).rotate(20)
mmi4.movey(50).movex(200).rotate(45)
mmi5.movey(200).movex(200).rotate(-20)

ports1 = [mmi1.ports["o2"], mmi2.ports["o1"]]
ports2 = [mmi3.ports["o1"], mmi4.ports["o1"]]
ports3 = [mmi3.ports["o2"], mmi5.ports["o2"]]
ports4 = [mmi5.ports["o1"], mmi1.ports["o3"]]

routes1 = gf.routing.route_bundle_all_angle(c, ports1, ports2)
routes2 = gf.routing.route_bundle_all_angle(c, ports3, ports4)

for inst in c.insts:
print(f"inst type: {type(inst)}")
try:
polygons_points = inst.cell.get_polygons_points()
print(list(polygons_points.values())[0][0])
except Exception as e:
print(f"error: {e}")
c.show()

Output：
%runfile 'C:/Users/lenovo/Desktop/Paper Code2/未命名1.py' --wdir
inst type: <class 'kfactory.kcell.VInstance'>
[[ 0. -1.25 ]
[ 0. -0.5 ]
[-10. -0.25 ]
[-10. 0.25 ]
[ 0. 0.5 ]
[ 0. 1.25 ]
[ 5.5 1.25 ]
[ 5.5 1.125]
[ 15.5 0.875]
[ 15.5 0.375]
[ 5.5 0.125]
[ 5.5 -0.125]
[ 15.5 -0.375]
[ 15.5 -0.875]
[ 5.5 -1.125]
[ 5.5 -1.25 ]]
inst type: <class 'kfactory.kcell.VInstance'>
[[ 0. -1.25 ]
[ 0. -0.5 ]
[-10. -0.25 ]
[-10. 0.25 ]
[ 0. 0.5 ]
[ 0. 1.25 ]
[ 5.5 1.25 ]
[ 5.5 1.125]
[ 15.5 0.875]
[ 15.5 0.375]
[ 5.5 0.125]
[ 5.5 -0.125]
[ 15.5 -0.375]
[ 15.5 -0.875]
[ 5.5 -1.125]
[ 5.5 -1.25 ]]
inst type: <class 'kfactory.kcell.VInstance'>
[[ 0. -1.25 ]
[ 0. -0.5 ]
[-10. -0.25 ]
[-10. 0.25 ]
[ 0. 0.5 ]
[ 0. 1.25 ]
[ 5.5 1.25 ]
[ 5.5 1.125]
[ 15.5 0.875]
[ 15.5 0.375]
[ 5.5 0.125]
[ 5.5 -0.125]
[ 15.5 -0.375]
[ 15.5 -0.875]
[ 5.5 -1.125]
[ 5.5 -1.25 ]]
inst type: <class 'kfactory.kcell.VInstance'>
[[ 0. -1.25 ]
[ 0. -0.5 ]
[-10. -0.25 ]
[-10. 0.25 ]
[ 0. 0.5 ]
[ 0. 1.25 ]
[ 5.5 1.25 ]
[ 5.5 1.125]
[ 15.5 0.875]
[ 15.5 0.375]
[ 5.5 0.125]
[ 5.5 -0.125]
[ 15.5 -0.375]
[ 15.5 -0.875]
[ 5.5 -1.125]
[ 5.5 -1.25 ]]
inst type: <class 'kfactory.kcell.VInstance'>
[[ 0. -1.25 ]
[ 0. -0.5 ]
[-10. -0.25 ]
[-10. 0.25 ]
[ 0. 0.5 ]
[ 0. 1.25 ]
[ 5.5 1.25 ]
[ 5.5 1.125]
[ 15.5 0.875]
[ 15.5 0.375]
[ 5.5 0.125]
[ 5.5 -0.125]
[ 15.5 -0.375]
[ 15.5 -0.875]
[ 5.5 -1.125]
[ 5.5 -1.25 ]]
inst type: <class 'kfactory.kcell.VInstance'>
error: 'ComponentAllAngle' object has no attribute 'cell'
错误：'ComponentAllAngle' 对象没有 'cell' 属性
inst type: <class 'kfactory.kcell.VInstance'>
error: 'ComponentAllAngle' object has no attribute 'cell'
错误：'ComponentAllAngle' 对象没有 'cell' 属性
inst type: <class 'kfactory.kcell.VInstance'>
error: 'ComponentAllAngle' object has no attribute 'cell'
错误：'ComponentAllAngle' 对象没有 'cell' 属性
inst type: <class 'kfactory.kcell.VInstance'>
error: 'ComponentAllAngle' object has no attribute 'cell'
错误：'ComponentAllAngle' 对象没有 'cell' 属性
inst type: <class 'kfactory.kcell.VInstance'>
error: list index out of range
错误：列表索引超出范围
inst type: <class 'kfactory.kcell.VInstance'>
error: 'ComponentAllAngle' object has no attribute 'cell'
错误：'ComponentAllAngle' 对象没有 'cell' 属性
inst type: <class 'kfactory.kcell.VInstance'>
error: 'ComponentAllAngle' object has no attribute 'cell'
错误：'ComponentAllAngle' 对象没有 'cell' 属性
inst type: <class 'kfactory.kcell.VInstance'>
error: 'ComponentAllAngle' object has no attribute 'cell'
错误：'ComponentAllAngle' 对象没有 'cell' 属性
inst type: <class 'kfactory.kcell.VInstance'>
error: list index out of range
inst type: <class 'kfactory.kcell.VInstance'>
error: 'ComponentAllAngle' object has no attribute 'cell'
inst type: <class 'kfactory.kcell.VInstance'>
error: 'ComponentAllAngle' object has no attribute 'cell'
inst type: <class 'kfactory.kcell.VInstance'>
error: 'ComponentAllAngle' object has no attribute 'cell'
inst type: <class 'kfactory.kcell.VInstance'>
error: 'ComponentAllAngle' object has no attribute 'cell'
inst type: <class 'kfactory.kcell.VInstance'>
error: 'ComponentAllAngle' object has no attribute 'cell'
inst type: <class 'kfactory.kcell.VInstance'>
error: 'ComponentAllAngle' object has no attribute 'cell'
inst type: <class 'kfactory.kcell.VInstance'>
error: 'ComponentAllAngle' object has no attribute 'cell'
inst type: <class 'kfactory.kcell.VInstance'>
error: list index out of range

**Comment by sebastian-goeldi** [comment]
> 1. Could you clarify the difference in how route_bundle_all_angle handles or generates VInstance objects internally?

I am not sure I follow, essentially it will just create VInstance objects (in this case of VKCell) and add them to `c`. These are just vehicles allowing offgrid placing and coordinates.

> 2. Is there a recommended approach to accessing polygon data from the OpticalAllAngleRoute objects returned by 

There is not. It's not currently really possible to access polygons of vinsts. There could be, but it's a lot of manual transformations.

You can insert them into a separate component though and use `separate_component.insert_vinsts()` to convert them to real instances and polygons, but it's also not ideal most likely.

**Reply by sebastian-goeldi** [reply]
Also, please don't open an issue and a discussion with the same text ;)

**Reply by Dangerance** [reply]
Thank you for your prompt reply and for clarifying how route_bundle_all_angle creates VInstance objects internally. I also appreciate the reminder about not opening both an issue and a discussion with the same text—my apologies for that. I understand now that VInstance objects are essentially vehicles for placing elements off-grid and that accessing their polygons is currently not straightforward.

I do have a couple of follow-up questions:

It seems that route_bundle_all_angle() modifies the input component c by inserting the routed waveguide paths, but it doesn’t return a modified version of c—it just returns the routes. If I understand correctly, c is updated in-place. Could you confirm whether c is indeed modified internally by the function, and if so, how is this achieved without explicitly assigning the result back to c?

You mentioned that it’s not currently possible to access polygon data directly from these VInstance objects. However, when I use gf.ComponentAllAngle() and then create_vinst() to generate, for example, MMI components, I’m able to generate polygons for them without issues. On the other hand, the routed waveguides inserted by route_bundle_all_angle()—which I believe should be similar VInstance objects—cause an attribute error ('ComponentAllAngle' object has no attribute 'cell'). Why does the same approach work for the create_vinst() MMIs but not for the waveguides generated by route_bundle_all_angle()? Is there a fundamental difference in how these instances are constructed or stored?How to solve the problem of function route_bundle_all_angle() to generate polygons?

Thank you again for your time and assistance. Any guidance or insights you can provide would be greatly appreciated.

**Reply by sebastian-goeldi** [reply]
I need to check the exact code in gdsfactory (I only know the backend code from kfactory very well).

Essentially, the routing function uses `create_vinst`. All the resulting vinsts will be in `c.vinsts`.

Additionally, all vinsts should be in the returned routing object(s), either as `.instances` or something similar (again can't remember exact implementation).

The big difference to an MMI is that the underlying cell the VInstance is pointing to is a `Component` (or in the backend a `KCell`), whereas the routings use virtual components/cells (`ComponentAllAngle` or in the backend `VKCell`). I think this is where the problem comes from, but I am not 100% certain.
